jpeg.go

```diff
@@ -5,49 +5,156 @@ import (
+	// the naive multireader will just give us one byte from the first reader in its arguments
+	if len(p) > 1 {
+		if n, err = io.ReadAtLeast(teedReader, p, 2); err != nil {
+			if errors.Is(err, io.EOF) {
+				return
+			} else if errors.Is(err, io.ErrUnexpectedEOF) {
+				// safe to just deal with on the next pass through
+				err = nil
+			} else {
+				err = &MalformedDataError{"unexpected error reading inside scan data", err}
+				return
+			}
+		}
+	} else {
+		// if len(p) is only 1, we can still attempt to read because we'll still
+		// peak the next byte and can act accordingly
+		if n, err = teedReader.Read(p); err != nil {
+			if errors.Is(err, io.EOF) {
+				return
+			}
+			err = &MalformedDataError{"unexpected error reading inside scan data", err}
+			return
+		}
+	}
+
+	// attempt to peek one more byte to ensure that the marker doesn't overlap
+	// the bytes returned
+	lastByte := make([]byte, 1)
+	var lastByteN int
+	if lastByteN, err = io.ReadFull(teedReader, lastByte); err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			// an EOF is ok here because we'll catch it normally on the next pass through
+			err = nil
+		} else {
+			err = &MalformedDataError{"unexpected error reading inside scan data", err}
+			return
+		}
+	}
+
+	var markerIndex int
+	if lastByteN == 1 {
+		markerIndex = firstMarker(append(p[:n], lastByte[0]))
+	} else {
+		markerIndex = firstMarker(p[:n])
+	}
+
+	if markerIndex >= 0 {
+		n = markerIndex
+		if _, err = io.CopyN(ioutil.Discard, buf, int64(markerIndex)); err != nil {
+			err = &MalformedDataError{"unexpected error reading inside scan data", err}
+			return
+		}
+
+		// pop "unread" bytes back onto reader
+		*sr.reader = io.MultiReader(buf, *sr.reader)
+		if markerIndex == 0 {
+			err = io.EOF
+		}
+		return
+	}
+
+	if lastByteN == 1 {
+		// pop the last byte back onto the reader
+		*sr.reader = io.MultiReader(bytes.NewReader(lastByte), *sr.reader)
+	}
+
+	return
+}
+
 type jpegSegmentReader struct {
-	reader io.Reader
+	reader     io.Reader
+	insideScan bool
+	endOfImage bool
 }
 
 func (jr *jpegSegmentReader) isMetadataType(segmentMarker []byte) bool {
 	// APP segments are of form 0xff, 0xe[0-f]
+	// more information about them at https://exiftool.org/TagNames/JPEG.html
+	// APP14 and APP15 *seem* to be required to actually decode image data
 	// COM segments are of form 0xff, 0xfe
 	return segmentMarker[0] == 0xff &&
-		((segmentMarker[1] >= 0xe0 && segmentMarker[1] <= 0xef) || segmentMarker[1] == 0xfe)
+		((segmentMarker[1] >= 0xe0 && segmentMarker[1] <= 0xed) || segmentMarker[1] == 0xfe)
 }
 
 func (jr *jpegSegmentReader) nextSegment() (r io.Reader, isMetadata bool, err error) {
+	if jr.endOfImage {
+		return nil, false, io.EOF
+	}
+
+	if jr.insideScan {
+		jr.insideScan = false
+		return &scanReader{&jr.reader}, false, nil
+	}
+
 	segmentHeaderBuffer := new(bytes.Buffer)
 	teedReader := io.TeeReader(jr.reader, segmentHeaderBuffer)
```

Comment:
I realize this is existing code, but the `bufio` package may help here too.

Reply:
I super wish that were true and I'd love to be proven wrong, but I just didn't have any luck w/ any of `bufio`'s tooling :/

Reply:
Ah okay, didn't realize you'd tried it!